river/tree/base.py

```diff
@@ -28,14 +28,29 @@ def __init__(self, *children):
     def next(self, x) -> Union["Branch", "Leaf"]:
         """Move to the next node down the tree."""
 
-    def walk(self, x) -> Iterable[Union["Branch", "Leaf"]]:
+    @abc.abstractmethod
+    def most_common_path(self) -> Tuple[int, Union["Leaf", "Branch"]]:
+        """Return a tuple with the branch index and the child node related to the most
+        traversed path.
+
+        Used in case the split feature is missing from an instance.
+        """
+        pass
+
+    def walk(self, x, until_leaf=True) -> Iterable[Union["Branch", "Leaf"]]:
         """Iterate over the nodes of the path induced by x."""
         yield self
-        yield from self.next(x).walk(x)
-
-    def traverse(self, x) -> "Leaf":
+        try:
+            yield from self.next(x).walk(x, until_leaf)
+        except KeyError:
+            if until_leaf:
+                _, node = self.most_common_path()
+                yield node
+                yield from node.walk(x, until_leaf)
```

Comment:
I'm not clear on what this is doing (then again, it's late). Can you remind me why you need to introduce `until_leaf`? Isn't this specific to a particular tree?

Reply:
Nope, it's not! We follow some conventions to deal with missing split features. When learning or predicting we pass the instance to the most traversed path.

Ok, so in these cases, we always need to reach a leaf. But, there are situations where we have categorical features (or discretized numerical features) and a new category arrives. In these cases, we only pass the instance to the most traversed path while predicting (because the branch does not know what to do with the new category). However, while learning, we need to make sure we create a new path corresponding to this new category.

Hence, depending on the situation, we allow instances to stop at a branch, rather than always reaching a leaf. That's what the `until_leaf` param is designed to control. All the decision trees make calls to `walk` and `traverse` using different values for this parameter. Anyway, its default behavior is to always reach a tree.

Reply:
Right, but as we discussed elsewhere (I can't remember haha), can't you just handle the new leaf creation in the `next` function?

Reply:
Please check my previous comment for more info. But in summary, it's not that simple. The trees create different types of leaves depending on the user's choice. Some of them are simple, like those that output the majority class or the target mean. Others perform fancy operations. Usually, to build a new leaf, we need information from the main tree structure (e.g., leaf type, prediction model, seed, etc.) and the previous leaf to be promoted to an internal node. We could init the branches with the needed parameters from the main tree, but that doesn't sound much future proof nor easy to document. For instance, this also doesn't solve the situation where we copy the linear model from the leaf to be split to all its children.